File: actions.ts
// src/actions.ts
import { gmail } from './auth';

export async function markEmailAsRead(emailId: string) {
  await gmail.users.messages.modify({
    userId: 'me',
    id: emailId,
    requestBody: {
      removeLabelIds: ['UNREAD'],
    },
  });
}

export async function applyLabel(emailId: string, labelName: string) {
  const labelRes = await gmail.users.labels.list({ userId: 'me' });
  const label = labelRes.data.labels?.find((l) => l.name === labelName);

  let labelId = label?.id;

  if (!labelId) {
    const createRes = await gmail.users.labels.create({
      userId: 'me',
      requestBody: {
        name: labelName,
        labelListVisibility: 'labelShow',
        messageListVisibility: 'show',
      },
    });
    labelId = createRes.data.id!;
  }

  await gmail.users.messages.modify({
    userId: 'me',
    id: emailId,
    requestBody: {
      addLabelIds: [labelId!],
    },
  });
}

File: app/actions.py
from typing import Any
from app.auth import get_gmail_service

def mark_email_as_read(email_id: str) -> None:
    """
    # ...existing code...
    """
    service = get_gmail_service()
    service.users().messages().modify(
        userId='me',
        id=email_id,
        body={'removeLabelIds': ['UNREAD']}
    ).execute()

def apply_label(email_id: str, label_name: str) -> None:
    """
    # ...existing code...
    """
    service = get_gmail_service()
    labels_res = service.users().labels().list(userId='me').execute()
    labels = labels_res.get('labels', [])
    label_id = None
    for label in labels:
        if label['name'] == label_name:
            label_id = label['id']
            break
    if not label_id:
        create_res = service.users().labels().create(
            userId='me',
            body={
                'name': label_name,
                'labelListVisibility': 'labelShow',
                'messageListVisibility': 'show',
            }
        ).execute()
        label_id = create_res['id']
    service.users().messages().modify(
        userId='me',
        id=email_id,
        body={'addLabelIds': [label_id]}
    ).execute()
